File: 2017/day/3/spiral-memory-sum.go
package main

import "fmt"
import "os"
import "strconv"


var LEFT = [2]int{-1, 0}
var RIGHT = [2]int{1, 0}
var UP = [2]int{0, 1}
var DOWN = [2]int{0, -1}

var UP_LEFT = [2]int{-1, 1}
var UP_RIGHT = [2]int{1, 1}
var DOWN_LEFT = [2]int{-1, -1}
var DOWN_RIGHT = [2]int{1, -1}

var CARDINALS = [4][2]int{RIGHT, UP, LEFT, DOWN}
var DIAGNOALS = [4][2]int{UP_LEFT, UP_RIGHT, DOWN_LEFT, DOWN_RIGHT}

var ADJACENTS = [8][2]int{RIGHT, UP, LEFT, DOWN, UP_LEFT, UP_RIGHT, DOWN_LEFT, DOWN_RIGHT}



func move(point [2]int, offset [2]int) [2]int {
    return [2]int{point[0] + offset[0], point[1] + offset[1]}
}


func getNeighborSum(point [2]int, values map[[2]int]int) int {
    sum := 0
    for _, offset := range ADJACENTS {
        sum += values[move(point, offset)]
    }
    return sum
}


func getFirstValueGreaterThan(target int) int {
    length := 1
    point := [2]int{0, 0}
    adjustment := false

    values := make(map[[2]int]int)
    values[[2]int{0, 0}] = 1

    for {
        for _, direction := range CARDINALS {
            for i := 0; i < length; i++ {
                point = move(point, direction)
                values[point] = getNeighborSum(point, values)
                if values[point] > target {
                    return values[point]
                }
            }

            if adjustment {
                length += 1
            }
            adjustment = !adjustment
        }
    }
}


func main() {
    target, err := strconv.Atoi(os.Args[1])
    if err != nil || target < 1 {
        fmt.Print("That's not a positive integer... " + os.Args[1])
    } else {
        fmt.Print(getFirstValueGreaterThan(target))
    }
}

File: 2017/03/spiral-memory-sum.py
import sys
import collections


LEFT = -1, 0
RIGHT = 1, 0
UP = 0, 1
DOWN = 0, -1

UP_LEFT = -1, 1
UP_RIGHT = 1, 1
DOWN_LEFT = -1, -1
DOWN_RIGHT = 1, -1

CARDINALS = [UP, DOWN, LEFT, RIGHT]
DIAGNOALS = [UP_LEFT, UP_RIGHT, DOWN_LEFT, DOWN_RIGHT]


values = collections.defaultdict(int)
values[0, 0] = 1


def move(point, offset):
    return point[0] + offset[0], point[1] + offset[1]


def get_neighbors(point):
    for offset in CARDINALS + DIAGNOALS:
        yield values[move(point, offset)]


def get_first_value_greater_than(target):
    length = 1
    point = 0, 0
    adjustment = False

    while True:
        for direction in RIGHT, UP, LEFT, DOWN:
            for i in range(length):
                point = move(point, direction)
                values[point] = value = sum(get_neighbors(point))
                if value > target:
                    return value

            length += adjustment
            adjustment = not adjustment


if __name__ == '__main__':
    try:
        target = int(sys.argv[1])
    except ValueError:
        target = 0
    if target < 1:
        print(f"That's not a positive integer... {sys.argv[1]}")
    else:
        print(get_first_value_greater_than(target))
